libs/providers/go-feature-flag-web/src/lib/go-feature-flag-web-provider.ts

```diff
@@ -0,0 +1,269 @@
+
+  metadata = {
+    name: GoFeatureFlagWebProvider.name,
+  };
+  events = new OpenFeatureEventEmitter();
+
+  // logger is the Open Feature logger to use
+  private _logger?: Logger;
+  // endpoint of your go-feature-flag relay proxy instance
+  private readonly _endpoint: string;
+  // timeout in millisecond before we consider the http request as a failure
+  private readonly _apiTimeout: number;
+
+  // initial delay in millisecond to wait before retrying to connect the websocket
+  private readonly _websocketRetryInitialDelay;
+  // multiplier of _websocketRetryInitialDelay after each failure
+  private readonly _websocketRetryDelayMultiplier;
+  // maximum number of retries
+  private readonly _websocketMaxRetries;
+
+  // _websocket is the reference to the websocket connection
+  private _websocket?: WebSocket;
+  // _flags is the in memory representation of all the flags.
+  private _flags: { [key: string]: ResolutionDetails<FlagValue> } = {};
+
+
+  constructor(options: GoFeatureFlagWebProviderOptions, logger?: Logger) {
+    this._logger = logger;
+    this._apiTimeout = options.apiTimeout || 0; // default is 0 = no timeout
+    this._endpoint = options.endpoint;
+    this._websocketRetryInitialDelay = options.websocketRetryInitialDelay || 100;
+    this._websocketRetryDelayMultiplier = options.websocketRetryDelayMultiplier || 2;
+    this._websocketMaxRetries = options.websocketMaxRetries || 10;
+    // Add API key to the headers
+    if (options.apiKey) {
+      axios.defaults.headers.common['Authorization'] = `Bearer ${options.apiKey}`;
+    }
+  }
+
+  async initialize(context: EvaluationContext): Promise<void> {
+    await this.fetchAll(context);
+    await this.connectWebsocket();
+    this._logger?.debug("go-feature-flag provider initialized")
+  }
+
+  /**
+   * connectWebsocket is starting the websocket and associate some handler
+   * to react if the state of the websocket change.
+   */
+  async connectWebsocket(): Promise<void> {
+    const wsURL = new URL(this._endpoint);
+    wsURL.pathname =
+      wsURL.pathname.endsWith('/') ? wsURL.pathname + this._websocketPath : wsURL.pathname + '/' + this._websocketPath;
+    wsURL.protocol = "ws"
+    this._logger?.debug(`Trying to connect the websocket at ${wsURL}`)
+
+    this._websocket = new WebSocket(wsURL, ["ws", "http", "https"]);
+    await this.waitWebsocketFinalStatus(this._websocket);
+
+    this._websocket.onopen = (event) => {
+      this._logger?.info(`Websocket to go-feature-flag open: ${event}`);
+    };
+    this._websocket.onmessage = async ({data}) => {
+      this._logger?.info(`Change in your configuration flag`);
+      const t: GOFeatureFlagWebsocketResponse = JSON.parse(data);
+      const flagsChanged = this.extractFlagNamesFromWebsocket(t);
+      await this.fetchAll(OpenFeature.getContext(), flagsChanged);
+    }
+    this._websocket.onclose = async () => {
+      this._logger?.info(`Websocket closed, trying to reconnect`);
```

Comment:
It could be a lot of things here, such as the connection is closed from the upstream, the server does not exist anymore because we have downscale etc ...

But I agree using `warn` here is better.